File: android/app/src/main/java/dev/yury/mathmark/Stats.kt
package dev.yury.mathmark

import java.io.File
import java.time.LocalDate
import java.time.LocalDateTime

/**
 * Журнал отметок и подсчёт движения по нему.
 *
 * Каждое нажатие на кружок дописывает одну строку в обычный текстовый файл
 * рядом с настройками:
 *
 *     2026-07-31T18:50:12|Матан.md|task|done
 *
 * Именно момент отметки, а не момент синхронизации: отметил утром,
 * синхронизировал через три дня — статистика должна показать утро.
 *
 * Файл только растёт и никогда не переписывается. Строку можно удалить
 * руками, ничего не сломается.
 *
 * Точный повтор `stats.py` настольной версии, вплоть до тех же проверок.
 */

data class JournalEntry(
    val whenAt: LocalDateTime,
    val path: String,
    val kind: Kind,
    val mark: Mark,
)

data class Stats(
    var todayTasks: Int = 0,
    var todayTopics: Int = 0,
    var weekTasks: Int = 0,
    var weekTopics: Int = 0,
    var monthTasks: Int = 0,
    var monthTopics: Int = 0,
    var streak: Int = 0,
    var perDay: List<Pair<LocalDate, Int>> = emptyList(),
)

object Journal {

    /** Дописать строку. Беда с записью не должна мешать работе — молчим. */
    fun record(file: File, path: String, kind: Kind, mark: Mark, at: LocalDateTime = LocalDateTime.now()) {
        val line = "${at.withNano(0)}|$path|${kind.name.lowercase()}|${mark.char}\n"
        runCatching {
            file.parentFile?.mkdirs()
            file.appendText(line, Charsets.UTF_8)
        }
    }

    fun parse(text: String): List<JournalEntry> {
        val out = ArrayList<JournalEntry>()
        for (raw in text.split("\n")) {
            val parts = raw.trim().split("|")
            if (parts.size != 4) continue
            val at = runCatching { LocalDateTime.parse(parts[0]) }.getOrNull() ?: continue
            out.add(
                JournalEntry(
                    whenAt = at,
                    path = parts[1],
                    kind = if (parts[2] == "task") Kind.TASK else Kind.TOPIC,
                    mark = Mark.of(parts[3].firstOrNull() ?: ' '),
                )
            )
        }
        return out
    }

    /**
     * Считаем только закрытия — переход в «готово». Снятие отметки не отнимает:
     * это не отчётность, а счётчик движения вперёд.
     */
    fun summarise(entries: List<JournalEntry>, today: LocalDate = LocalDate.now()): Stats {
        val s = Stats()
        val byDay = HashMap<LocalDate, Int>()

        for (e in entries) {
            if (e.mark != Mark.DONE) continue
            val d = e.whenAt.toLocalDate()
            byDay[d] = (byDay[d] ?: 0) + 1
            val days = java.time.temporal.ChronoUnit.DAYS.between(d, today).toInt()
            if (days == 0) {
                if (e.kind == Kind.TASK) s.todayTasks++ else s.todayTopics++
            }
            if (days in 0..6) {
                if (e.kind == Kind.TASK) s.weekTasks++ else s.weekTopics++
            }
            if (days in 0..29) {
                if (e.kind == Kind.TASK) s.monthTasks++ else s.monthTopics++
            }
        }

        var day = today
        while ((byDay[day] ?: 0) > 0) {
            s.streak++
            day = day.minusDays(1)
        }

        s.perDay = (29 downTo 0).map { i ->
            val d = today.minusDays(i.toLong())
            d to (byDay[d] ?: 0)
        }
        return s
    }
}

File: desktop/mathmark/stats.py
"""
Журнал отметок и подсчёт движения по нему.

Каждое нажатие на кружок дописывает одну строку в обычный текстовый файл
рядом с настройками:

    2026-07-31T18:50:12|Матан.md|task|done

Именно момент отметки, а не момент синхронизации: отметил утром,
синхронизировал через три дня — статистика должна показать утро.

Файл только растёт и никогда не переписывается, поэтому испортить его
нечем. Строку можно удалить руками — ничего не сломается.
"""

from __future__ import annotations

from collections import Counter
from dataclasses import dataclass, field
from datetime import date, datetime, timedelta
from pathlib import Path

from .md_items import Kind, Mark


@dataclass
class Entry:
    when: datetime
    path: str
    kind: Kind
    mark: Mark


@dataclass
class Stats:
    """Сводка для экрана «сколько сделано»."""
    today_tasks: int = 0
    today_topics: int = 0
    week_tasks: int = 0
    week_topics: int = 0
    month_tasks: int = 0
    month_topics: int = 0
    streak: int = 0                       # дней подряд, считая назад от сегодня
    per_day: list[tuple[date, int]] = field(default_factory=list)   # последние 30 дней


def record(journal: Path, path: str, kind: Kind, mark: Mark, when: datetime | None = None) -> None:
    """Дописать строку. Ошибка записи не должна мешать работе — молчим."""
    when = when or datetime.now()
    line = f"{when.isoformat(timespec='seconds')}|{path}|{kind.value}|{mark.value}\n"
    try:
        journal.parent.mkdir(parents=True, exist_ok=True)
        with journal.open("a", encoding="utf-8") as fh:
            fh.write(line)
    except OSError:
        pass


def parse(text: str) -> list[Entry]:
    out: list[Entry] = []
    for raw in text.split("\n"):
        parts = raw.strip().split("|")
        if len(parts) != 4:
            continue
        try:
            when = datetime.fromisoformat(parts[0])
        except ValueError:
            continue
        kind = Kind.TASK if parts[2] == "task" else Kind.TOPIC
        out.append(Entry(when, parts[1], kind, Mark.of(parts[3] or " ")))
    return out


def summarise(entries: list[Entry], today: date | None = None) -> Stats:
    """
    Считаем только закрытия — переход в «готово». Снятие отметки не отнимает:
    это не отчётность, а счётчик движения вперёд.
    """
    today = today or date.today()
    done = [e for e in entries if e.mark is Mark.DONE]
    s = Stats()

    by_day: Counter[date] = Counter()
    for e in done:
        d = e.when.date()
        by_day[d] += 1
        days = (today - d).days
        if days == 0:
            if e.kind is Kind.TASK:
                s.today_tasks += 1
            else:
                s.today_topics += 1
        if 0 <= days < 7:
            if e.kind is Kind.TASK:
                s.week_tasks += 1
            else:
                s.week_topics += 1
        if 0 <= days < 30:
            if e.kind is Kind.TASK:
                s.month_tasks += 1
            else:
                s.month_topics += 1

    # серия: сколько дней подряд подряд что-то закрывалось, считая назад
    day = today
    while by_day.get(day, 0) > 0:
        s.streak += 1
        day -= timedelta(days=1)

    s.per_day = [(today - timedelta(days=i), by_day.get(today - timedelta(days=i), 0))
                 for i in range(29, -1, -1)]
    return s
